File: examples/quick-git/model/watcher.ts
import { FileWatcher } from "@quickgui/native";
import { relative, sep } from "node:path";

export type ChangeKind = "worktree" | "refs" | "index";

export interface RepositoryWatcherOptions {
  /** Working-tree root. */
  root: string;
  /** This worktree's own git directory (`.git` or `.git/worktrees/<name>`). */
  gitDir: string;
  /** Shared git directory, watched separately when it lies outside `root`. */
  commonDir: string;
  /** Called once per quiet period with every kind of change seen in it. */
  onChange: (kinds: ReadonlySet<ChangeKind>) => void;
  debounceMs?: number;
}

/**
 * Watch a working tree and its git directory, coalescing bursts into one callback.
 *
 * git writes lock files, temporary objects, and the index in quick succession; a 150 ms quiet
 * period folds a whole `git commit` into one refresh. Object writes are ignored outright: they
 * never change what the app shows without a ref or index change following them.
 */
export function watchRepository(options: RepositoryWatcherOptions): () => void {
  const debounceMs = options.debounceMs ?? 150;
  const pending = new Set<ChangeKind>();
  let timer: ReturnType<typeof setTimeout> | undefined;
  let closed = false;
  const flush = () => {
    timer = undefined;
    if (closed || pending.size === 0) return;
    const kinds = new Set([...pending]);
    pending.clear();
    options.onChange(kinds);
  };
  const note = (kind: ChangeKind) => {
    if (closed) return;
    pending.add(kind);
    if (timer) clearTimeout(timer);
    timer = setTimeout(flush, debounceMs);
  };
  let watcher: FileWatcher | undefined;
  const roots = [options.root];
  if (!isInside(options.gitDir, options.root)) roots.push(options.gitDir);
  if (options.commonDir !== options.gitDir && !isInside(options.commonDir, options.root)) roots.push(options.commonDir);
  const start = async (): Promise<void> => {
    try {
      const active = await FileWatcher.start(roots, (event) => {
        if (event.rescan) { note("worktree"); note("refs"); note("index"); }
        for (const path of event.paths) {
          const kind = isInside(path, options.gitDir)
            ? classifyGitPath(relative(options.gitDir, path))
            : isInside(path, options.commonDir)
              ? classifyGitPath(relative(options.commonDir, path))
              : classifyWorktreePath(relative(options.root, path), relative(options.root, options.gitDir));
          if (kind !== undefined) note(kind);
        }
      });
      if (closed) await active.close();
      else watcher = active;
    } catch (error) {
      // Focus refresh and the refresh after each write remain available if watching fails.
      console.error("Unable to watch repository", String(error));
    }
  };
  void start();

  return () => {
    closed = true;
    if (timer) clearTimeout(timer);
    const active = watcher;
    if (active !== undefined) void active.close();
  };
}

function isInside(path: string, directory: string): boolean {
  const rel = relative(directory, path);
  return rel === "" || (!rel.startsWith("..") && !rel.startsWith(sep));
}

/** Classify a path relative to the working-tree root, where `.git` may be a directory or a file. */
export function classifyWorktreePath(path: string, gitDirRelative: string): ChangeKind | undefined {
  const normalized = path.split(sep).join("/");
  const gitRelative = gitDirRelative.split(sep).join("/");
  if (normalized === ".git" || normalized === gitRelative) return "refs";
  if (normalized.startsWith(".git/")) return classifyGitPath(normalized.slice(5));
  if (gitRelative && normalized.startsWith(`${gitRelative}/`)) return classifyGitPath(normalized.slice(gitRelative.length + 1));
  return "worktree";
}

/** Classify a path relative to a git directory. */
export function classifyGitPath(path: string): ChangeKind | undefined {
  const normalized = path.split(sep).join("/");
  if (normalized.startsWith("objects/") || normalized.startsWith("lfs/") || normalized === "objects") return undefined;
  if (normalized === "index" || normalized === "index.lock") return "index";
  if (
    normalized === "HEAD" ||
    normalized === "ORIG_HEAD" ||
    normalized === "FETCH_HEAD" ||
    normalized === "MERGE_HEAD" ||
    normalized === "packed-refs" ||
    normalized.startsWith("refs/") ||
    normalized.startsWith("logs/") ||
    normalized.startsWith("worktrees/")
  ) {
    return "refs";
  }
  if (normalized.endsWith(".lock") || normalized === "COMMIT_EDITMSG") return undefined;
  return "refs";
}

File: examples/quick-git/internal/model/watcher.go
package model

import (
	"log"
	"path/filepath"
	"strings"
	"time"

	"github.com/egoist/quickgui/go/native"
)

type ChangeKind string

const (
	ChangeWorktree ChangeKind = "worktree"
	ChangeRefs     ChangeKind = "refs"
	ChangeIndex    ChangeKind = "index"
)

type WatcherOptions struct {
	Root      string
	GitDir    string
	CommonDir string
	OnChange  func(kinds map[ChangeKind]struct{})
	Debounce  time.Duration
}

func ClassifyGitPath(path string) ChangeKind {
	normalized := filepath.ToSlash(path)
	if strings.HasPrefix(normalized, "objects/") || strings.HasPrefix(normalized, "lfs/") || normalized == "objects" {
		return ""
	}
	if normalized == "index" || normalized == "index.lock" {
		return ChangeIndex
	}
	if normalized == "HEAD" || normalized == "ORIG_HEAD" || normalized == "FETCH_HEAD" ||
		normalized == "MERGE_HEAD" || normalized == "packed-refs" ||
		strings.HasPrefix(normalized, "refs/") || strings.HasPrefix(normalized, "logs/") ||
		strings.HasPrefix(normalized, "worktrees/") {
		return ChangeRefs
	}
	if strings.HasSuffix(normalized, ".lock") || normalized == "COMMIT_EDITMSG" {
		return ""
	}
	return ChangeRefs
}

func ClassifyWorktreePath(path, gitDirRelative string) ChangeKind {
	normalized := filepath.ToSlash(path)
	gitRelative := filepath.ToSlash(gitDirRelative)
	if normalized == ".git" || normalized == gitRelative {
		return ChangeRefs
	}
	if strings.HasPrefix(normalized, ".git/") {
		return ClassifyGitPath(normalized[5:])
	}
	if gitRelative != "" && strings.HasPrefix(normalized, gitRelative+"/") {
		return ClassifyGitPath(normalized[len(gitRelative)+1:])
	}
	return ChangeWorktree
}

// WatchRepository uses the Rust core's operating-system watcher. Debouncing
// starts only after a notification, so an unchanged repository has no polling loop.
func WatchRepository(options WatcherOptions) func() {
	debounce := options.Debounce
	if debounce <= 0 {
		debounce = 150 * time.Millisecond
	}
	pending := map[ChangeKind]struct{}{}
	var timer *time.Timer
	closed := false
	flush := func() {
		if closed || len(pending) == 0 {
			return
		}
		kinds := pending
		pending = map[ChangeKind]struct{}{}
		timer = nil
		if options.OnChange != nil {
			options.OnChange(kinds)
		}
	}
	note := func(kind ChangeKind) {
		if kind == "" {
			return
		}
		pending[kind] = struct{}{}
		if timer != nil {
			timer.Stop()
		}
		timer = time.AfterFunc(debounce, func() { native.Dispatch(flush) })
	}
	paths := []string{options.Root}
	for _, path := range []string{options.GitDir, options.CommonDir} {
		if path == "" {
			continue
		}
		exists := false
		for _, previous := range paths {
			if previous == path {
				exists = true
			}
		}
		if !exists {
			paths = append(paths, path)
		}
	}
	stop := native.WatchFiles(
		paths,
		func(event native.FileWatchEvent) {
			if closed {
				return
			}
			if event.Error != "" {
				log.Printf("Quick Git watcher: %s", event.Error)
			}
			if event.Rescan || event.Error != "" {
				note(ChangeWorktree)
				note(ChangeRefs)
				note(ChangeIndex)
			}
			for _, path := range event.Paths {
				gitPath := false
				for _, directory := range []string{options.GitDir, options.CommonDir} {
					if relative, ok := relativeChild(directory, path); ok {
						note(ClassifyGitPath(relative))
						gitPath = true
						break
					}
				}
				if gitPath {
					continue
				}
				if relative, ok := relativeChild(options.Root, path); ok {
					ignored := false
					for _, segment := range strings.Split(filepath.ToSlash(relative), "/") {
						if segment == "node_modules" || segment == ".quickgui" || segment == "target" {
							ignored = true
							break
						}
					}
					if !ignored {
						note(ClassifyWorktreePath(relative, ".git"))
					}
				}
			}
		},
		func(err error) {
			if err != nil {
				log.Printf("Quick Git watcher: %v", err)
			}
		},
	)
	return func() {
		if closed {
			return
		}
		closed = true
		if timer != nil {
			timer.Stop()
		}
		stop()
	}
}

func relativeChild(directory, path string) (string, bool) {
	if directory == "" {
		return "", false
	}
	relative, err := filepath.Rel(directory, path)
	return relative, err == nil && relative != ".." && !strings.HasPrefix(relative, ".."+string(filepath.Separator))
}
